internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -1551,3 +1616,100 @@ func (r *BMACReconciler) configureBMHForCleaning(ctx context.Context, log logrus
+
+	if !startExists || !timeoutExists {
+		return false, nil
+	}
+
+	drainTimeout, err := time.ParseDuration(timeoutValue)
+	if err != nil {
+		return false, err
+	}
+	firstTimeDrain, err := time.Parse(time.RFC3339, startValue)
+	if err != nil {
+		return false, err
+	}
+	now := time.Now().UTC()
+	diff := now.Sub(firstTimeDrain)
+	return diff.Seconds() >= drainTimeout.Seconds(), nil
+}
+
+func nodeUnreachable(node *corev1.Node) bool {
+	if node == nil {
+		return false
+	}
+	for _, c := range node.Status.Conditions {
+		if c.Type == corev1.NodeReady {
+			return c.Status == corev1.ConditionUnknown
+		}
+	}
+	return false
+}
+
+// Mostly taken from https://github.com/kubernetes-sigs/cluster-api/blob/539760a/internal/controllers/machine/machine_controller.go#L586
+func (r *BMACReconciler) drainAgentNode(ctx context.Context, log logrus.FieldLogger, agent *aiv1beta1.Agent) (bool, error) {
+	log = log.WithFields(logrus.Fields{
+		"spoke_cluster_name":      agent.Spec.ClusterDeploymentName.Name,
+		"spoke_cluster_namespace": agent.Spec.ClusterDeploymentName.Namespace,
+	})
+	// get spoke connection
+	spokeSecret, err := spokeKubeconfigSecret(ctx, log, r.Client, r.APIReader, agent.Spec.ClusterDeploymentName)
+	if err != nil {
+		log.WithError(err).Error("failed to get spoke kubeconfig secret")
+		return false, err
+	}
+
+	client, clientset, err := r.SpokeK8sClientFactory.ClientAndSetFromSecret(spokeSecret)
+	if err != nil {
+		log.WithError(err).Error("failed to create spoke client")
+		return false, err
+	}
+
+	nodeName := getAgentHostname(agent)
+	node, err := client.GetNode(ctx, nodeName)
+	if err != nil {
+		log.WithError(err).Errorf("failed to get node %s", nodeName)
+		return false, err
+	}
+
+	out := new(bytes.Buffer)
+	drainHelper := &drain.Helper{
+		Client:              clientset,
+		Ctx:                 ctx,
+		Force:               true,
+		IgnoreAllDaemonSets: true,
+		DeleteEmptyDirData:  true,
+		GracePeriodSeconds:  -1,
+		Timeout:             20 * time.Second,
+		OnPodDeletedOrEvicted: func(pod *corev1.Pod, usingEviction bool) {
+			verbStr := "Deleted"
+			if usingEviction {
+				verbStr = "Evicted"
+			}
+			log.Infof("%s Pod %s/%s from Node %s", verbStr, pod.Namespace, pod.Name, nodeName)
+		},
+		Out:    out,
+		ErrOut: out,
+	}
+	if nodeUnreachable(node) {
+		// When the node is unreachable and some pods are not evicted for as long as this timeout, we ignore them.
+		drainHelper.SkipWaitForDeleteTimeoutSeconds = 60 * 5 // 5 minutes
+	}
+	if err := r.Drainer.RunCordonOrUncordon(drainHelper, node, true); err != nil {
+		log.WithError(err).Errorf("failed to cordon node %s: output: %s", nodeName, out)
+		return false, errors.Wrapf(err, "failed to cordon node %s", nodeName)
+	}
+	if err := r.Drainer.RunNodeDrain(drainHelper, nodeName); err != nil {
+		log.WithError(err).Warnf("failed to drain node %s within timeout", nodeName)
+		log.Debugf("node %s drain output: %s", nodeName, out)
+		return true, nil
+	}
+
+	return false, nil
```

Comment:
Return the error here and you can just eliminate the bool from the return, no?

Reply:
No, because we need to differentiate between three cases:
1. Some error happened that caused us not to attempt a drain (failed to connect, couldn't get node, etc)
    - Don't requeue the reconcile from the caller and return an error
    - `return false, error`
2. Drain ran but errored (timeout or otherwise)
    - Requeue the reconcile from the caller until we hit the timeout
    - `return true, nil`
3. Drain ran and succeeded
    - Continue the reconcile
    - `return false, nil`